### ISSUE TITLE
Question : How to apply style to a dragged item ?

### ISSUE THREAD

**Issue by RobbWatershed** [mikepenz/FastAdapter#886]
Heya,

I'm using v5.0.0. My users have a hard time seeing when an draggable list item enters the "dragged" state because it doesn't change visually.

=> I'd like to style (apply elevation and/or change the background color, to be precise) a dragged item, so that its new state becomes apparent.

Thing is, I read the sample activity and search closed issues, but I couldn't find a way to do that.

Any tip ?

Thanks in advance,

-Robb

## Checklist

- [x] Searched for [similar issues](https://github.com/mikepenz/FastAdapter/issues)
- [x] Checked out the [sample application](https://github.com/mikepenz/FastAdapter/tree/develop/app)
- [x] Read the [README](https://github.com/mikepenz/FastAdapter/blob/develop/README.md)
- [x] Checked out the [CHANGELOG](https://github.com/mikepenz/FastAdapter/releases)
- [x] Checked out the [MIGRATION GUIDE](https://github.com/mikepenz/FastAdapter/blob/develop/MIGRATION.md)

**Comment by mikepenz** [comment]
@RobbWatershed this is not really a part this library is responsible for. the FastAdapter provides the framework to build out lists and hook up various things. 

Styling. how things look... are all in your hand :) 

https://github.com/mikepenz/FastAdapter/blob/develop/app/src/main/res/layout/swipeable_item.xml#L7

**Comment by RobbWatershed** [comment]
@mikepenz thanks a lot for the quick reply and the link.

Now that's interesting : the link describes an item changing its aspect when _swiping_ starts. This is all fine by me, but what I need is the exact same thing for _dragging_.

However, afaik, the only callbacks I can use for that are :
- `ItemTouchCallback.itemTouchDropped` -> obviously too late to detect an item being dragged

- `ItemTouchCallback.itemTouchOnMove` -> too late as well, since it only triggers when the position of the dragged item is swapped with the position of another item (see attached video)

[20200428_133141.zip](https://github.com/mikepenz/FastAdapter/files/4545404/20200428_133141.zip)

=> Would it be possible to create a new callback on `ItemTouchCallback` that would be triggered _as soon as the item enters the dragged state_ (i.e. right after the long click action) ? 

Correct me if I'm wrong, but I have the feeling it would be too hacky to listen to the viewholder's root `View`'s `OnLongClickListener` to detect that...

**Comment by mikepenz** [comment]
You mean you want to modify the ui while the action is happening? 
(What I mainly was thinking that you want to adjust the UI of the left and right side) 

I believe without any change to the SDK you have 2 options. either you do the item touch code manually (without the simplified helper classes) 
or you use a long click event hoook on the item or even better a touch event hook so you directly get the event when the click state is over


I'll have to further look into this to consider adding a new callback method.

**Comment by RobbWatershed** [comment]
> You mean you want to modify the ui while the action is happening?

Asolutely. Take a look at the below video (source : https://material.io/components/lists#behavior). 

Material design guidelines indicate that the dragged item should change its appearance while being dragged. Here we can see it becomes slightly darker and gets elevation.

https://kstatic.googleusercontent.com/files/3b4f47c77955908daec564c62deebb0a1ea2e1dab3cd664580613d4e86675780f36b558adc40d13758e6fa0f8e505398e80ec51f922e1177d0447727077b63f6

I'll explore the options you just gave and keep you informed. It would be neat nevertheless to have a proper callback - the equivalent to `itemTouchDropped`, but for the beginning of the action 😉

**Comment by RobbWatershed** [comment]
Turns out option 2 (using `MotionEvent`s or `GestureDetector`) is not a good idea : the long click, up and cancel events don't always match the actual behaviour of FastAdapter, and I'm slowly starting to look at `ItemTouchHelper` to understand what's going on, which is a really bad sign, as I'll end up emulating your code with my code :-O

Gotta try option 1 tomorrow, but I fear I'm gonna need to write kilometers of code for one small esthetic need.

**Comment by RobbWatershed** [comment]
Alright, I've come to an acceptable solution.

I'd gladly suggest a PR but my AS tells me you're using AS 4.1 when trying to build the latest version, and I don't really want to use Canary.

1/ Extend `SimpleDragCallback` and override `onSelectedChanged` to give its caller the ability to detect start drag events (called "select" events by Android)

```java
public class CustomDragCallback extends SimpleDragCallback {

    private final Consumer<RecyclerView.ViewHolder> onStartDrag;

    public CustomDragCallback(
            int directions,
            ItemTouchCallback touchCallback,
            @NonNull Consumer<RecyclerView.ViewHolder> onStartDrag) {
        super(directions, touchCallback);
        this.onStartDrag = onStartDrag;
    }

    @Override
    public void onSelectedChanged(@Nullable RecyclerView.ViewHolder viewHolder, int actionState) {
        super.onSelectedChanged(viewHolder, actionState);
        if (ItemTouchHelper.ACTION_STATE_DRAG == actionState) onStartDrag.accept(viewHolder);
    }
}
```
NB : In an ideal world, these changes would be integrated directly into `SimpleDragCallback` 😉 

***

2/ Create `onStartDrag` / `onDropped` methods on the `ViewHolder` itself to update its contents. 

```java
public void onDragged() {
   rootView.setBackgroundColor(AAA);
}

public void onDropped() {
   rootView.setBackgroundColor(BBB);
}
```

***

3/ Plug both in the `Fragment` or `Activity`

```java
@Override
    public void itemTouchDropped(int oldPosition, int newPosition) {
        fastAdapter.notifyItemChanged();
        RecyclerView.ViewHolder vh = recyclerView.findViewHolderForAdapterPosition(newPosition);
        if (vh instanceof SiteItem.SiteViewHolder) { // <-- that's kinda lame; I might use an interface in the future
            ((SiteItem.SiteViewHolder) vh).onDropped();
        }
    }

    private void onStartDrag(RecyclerView.ViewHolder vh) {
        if (vh instanceof SiteItem.SiteViewHolder) {
            ((SiteItem.SiteViewHolder) vh).onDragged();
        }
    }
```

NB : Passing through an `AbstractItem` property and using `FastAdapter.notifyItemChanged` seems to interrupt the drag action, as the `RecyclerView` takes over to update the item. That's why I'm using the `ViewHolder` directly.

**Comment by mikepenz** [comment]
@RobbWatershed you can always install canary along (side by side) with stable. but yeah I understand. if you want to code locally you can downgrade the build gradle version to the version applicable for you. I mainly use canary for the great new goodies there are :) https://developer.android.com/studio/preview/features


> NB : Passing through an AbstractItem property and using FastAdapter.notifyItemChanged seems to interrupt the drag action, as the RecyclerView takes over to update the item. That's why I'm using the ViewHolder directly. 
Yeah I think in this case going via the VH is appropriate. 


Thank you very much for the code. I believe if we can take care of backwards compatibility then I am fine with adding the new method to overwrite, to provide easier access out of the box

**Comment by RobbWatershed** [comment]
If by "backward compatibility" you mean the use of `Consumer`, then the PR I just made should be fine.

Thanks to you I just discovered Canary can be installed along the stable release !

**Comment by mikepenz** [comment]
> Thanks to you I just discovered Canary can be installed along the stable release ! 

You are welcome :D. I can't remember the last time I was not on a canary :D 

-----

Thanks for the PR. please have a look at the comment :)

**Comment by mikepenz** [comment]
Merged your PR released with v5.0.1 :) 

Thanks a lot again
